diff --git a/package/base-files/files/etc/shadow b/package/base-files/files/etc/shadow
--- a/package/base-files/files/etc/shadow
+++ b/package/base-files/files/etc/shadow
+++ /tmp/local_shadow	2026-09-24 09:40:18.352000000 +0800
@@ -1,4 +1,5 @@
-root:::0:99999:7:::
+root:$1$yLLV5smu$orGjQCvzqcOAdCYuqY8dZ0:0:0:99999:7:::
 daemon:*:0:0:99999:7:::
+ftp:*:0:0:99999:7:::
 network:*:0:0:99999:7:::
 nobody:*:0:0:99999:7:::
